src/sig_stfl/sig_stfl.h

```diff
@@ -469,105 +483,118 @@ OQS_API void OQS_SIG_STFL_free(OQS_SIG_STFL *sig);
+ * This allows for proper synchronization in a multi-threaded or multi-process environment,
+ * by ensuring that a secret key is not used concurrently by multiple entities, which could otherwise lead to security issues.
  *
- * @param[in] sk secret key pointer to be updated
- * @param[in] lock function pointer
+ * @param[in] sk Pointer to the secret key object whose lock function is to be set.
+ * @param[in] lock Function pointer to the locking routine provided by the application.
+ * @return None.
  *
  */
 OQS_API void OQS_SIG_STFL_SECRET_KEY_SET_lock(OQS_SIG_STFL_SECRET_KEY *sk, lock_key lock);
 
 /**
- * OQS_SIG_STFL_SECRET_KEY_SET_unlock .
+ * Attach an unlock mechanism to a secret key object.
  *
- * Sets function to prevent multiple processes from using the sk at the same time.
+ * This allows for proper synchronization * in a multi-threaded or multi-process environment,
+ * by ensuring that a secret key is not used concurrently by multiple entities, which could otherwise lead to security issues.
  *
- * @param[in] sk secret key pointer to be updated
- * @param[in] unlock function pointer
+ * @param[in] sk Pointer to the secret key object whose unlock function is to be set.
+ * @param[in] unlock Function pointer to the unlock routine provided by the application.
+ * @return None.
  *
  */
 OQS_API void OQS_SIG_STFL_SECRET_KEY_SET_unlock(OQS_SIG_STFL_SECRET_KEY *sk, unlock_key unlock);
 
 /**
- * OQS_SIG_STFL_SECRET_KEY_SET_mutex .
+ * Assign a mutex function to handle concurrency control over the secret key.
  *
- * Sets function to prevent multiple processes from using the sk at the same time.
+ * This is to ensure that only one process can access or modify the key at any given time.
  *
- * @param[in] sk secret key pointer to be updated
- * @param[in] mutex function pointer
+ * @param[in] sk A pointer to the secret key that the mutex functionality will protect.
+ * @param[in] mutex A function pointer to the desired concurrency control mechanism.
+ * @return None.
  *
  */
 OQS_API void OQS_SIG_STFL_SECRET_KEY_SET_mutex(OQS_SIG_STFL_SECRET_KEY *sk, void *mutex);
 
 /**
- * OQS_SIG_STFL_SECRET_KEY_lock .
- *
- * Locks the secret key so only one application that holds the lock can access it.
+ * Lock the secret key to ensure exclusive access in a concurrent environment.
  *
- * @param[in] sk secret key pointer to be locked
- * @return OQS_SUCCESS if successful, or OQS_ERROR if the object fails to apply the lock
+ * This operation is essential in multi-threaded or multi-process contexts
+ * in order to prevent simultaneous operations that could compromise the stateful signature security.
  *
+ * @param[in] sk Pointer to the secret key to be locked.
+ * @return OQS_SUCCESS if the lock is successfully applied; OQS_ERROR otherwise.
  */
 OQS_API OQS_STATUS OQS_SIG_STFL_SECRET_KEY_lock(OQS_SIG_STFL_SECRET_KEY *sk);
 
 /**
- * OQS_SIG_STFL_SECRET_KEY_unlock .
+ * Unlock the secret key, making it accessible to other processes.
```

Comment:
Same question as above: When does (an application) need to call this function?

Reply:
So I added: 
```
* This operation is essential in multi-threaded or multi-process contexts
 * in order to prevent simultaneous operations that could compromise the stateful signature security.
 
 ...
* @note It's not required to use lock and unlock functions in a single-threaded environment.

```
So, it's should be call in multi-threaded environment. 
I wonder if the current documentation is good enough, let me know if it's unclear.

Reply:
> I wonder if the current documentation is good enough, let me know if it's unclear.

I can only judge from my perspective and cannot know what other's may (mis)understand, so here goes my question now: Which (`liboqs` API) calls ("operations" in your wording above) need to be guarded with this function in multi-threaded environments? And a much more basic question: If there are such library functions failing to properly operate in a multi-threaded environment, wouldn't it be better to document _them_ accordingly? And as follow-on question: Does it need such `liboqs`-based "lock/unlock" functions at all if such documentation exists? Why isn't use of any external, e.g. as provided by the application-chosen threading library, lock/unlock mechanisms sufficient?

I guess what I'd like to suggest in a more "brutal question": Should we not strive to implement a thread-safe library API OR completely get out of the way (giving the app developer full control of what they want/need to do)? Anything in between IMO has the potential for a "support nightmare" ("use API XYZ if you are using threading library ABC; if not using any threading, disregard XYZ but use API XZZ", etc....)

Reply:
I answer each question below. 

1. What function need to be guarded with this function in a concurrent environment? 

Only the Signing operation.

Thanks to your question, I updated this note: 

```
* @note It's not necessary to use this function in either Keygen or Verifying operations.
 *       In a concurrent environment, the user is responsible for locking and unlocking the private key,
 *       to make sure that only one thread can access the private key during a Signing operation.
 ```

2. Why such lock/unlock APIs are there? Is it meant we strive to be a thread-safe library?

These APIs intention is to limit to one owner of private key at a time in a concurrent environment.
If users supply lock/unlock logic with a mutex, we acquire the lock to be the only owner of the private key, and unlock after the Signing operation complete. 
If users don't  supply the lock/unlock logic or a mutex, they are simply NOP operations. 

I think the key different here is: We don't execute stateful signature in multi-thread  at the library itself, but we provide a way to let users use multi-thread securely, in the right way. 

3. Does it need such liboqs-based "lock/unlock" functions at all if such documentation exists? 

I think to have `lock/unlock` functions help liboqs in:

- Testing. Someone has to foreseen the usage of our proposed APIs in multi-thread environment, so we need `lock/unlock` to test if it's still safe in such if our low-level implementation change. 

So we have it in test, well, why not exposed them to library user, it's convenience anyway. So there are `liboqs`-based `lock/unlock` functions.

4. Why isn't use of any external, e.g. as provided by the application-chosen threading library, lock/unlock mechanisms sufficient?

Actually we don't concretely define the `lock/unlock` mechanisms in `liboqs`, if the function pointers of `lock/unlock` are non-NULL then we will run them to perform `lock/unlock` the private key. 

So, we open to any application-chosen threading library, and it's user responsibility to create a proper `lock/unlock` functions, and it's up to them to define their `mutex`. 

5. Anything in between IMO has the potential for a "support nightmare" ("use API XYZ if you are using threading library ABC; if not using any threading, disregard XYZ but use API XZZ", etc....)

I totally agree with you, we don't want to get into "support nightmare". I want to run away from it as far as possible. :)) 

You have more experience than me in this matter, please let me know if there is something I need to change to get out of "support nightmare".

Reply:
@baentsch  before I merge this PR, do I answer your questions?

Reply:
> You have more experience than me in this matter, please let me know if there is something I need to change to get out of "support nightmare".

I'm not sure I do -- I did not use these APIs -- yet. My "general" question regarding minimizing support effort is: Could we completely do away with those functions in the external OQS API? If we could, we won't have to support them.

> So we have it in test, well, why not exposed them to library user, it's convenience anyway. So there are liboqs-based lock/unlock functions.

That's what I'm concerned with: If it's mere "convenience" (for us) we shouldn't externally make it available. Any API externally visible creates more support effort. Please see the discussion in #1648 and decide whether mutex/lock/unlock is another candidate for this internal API.

Reply:
I double check, these APIs just for testing purpose, so I moved them to internal APIs. 
Updated in https://github.com/open-quantum-safe/liboqs/pull/1655/commits/8b0e15e5ca0772c958a4ac368a8cf0a13d63b70a

Reply:
The `set_lock, set_mutex, and set_unlock` should still be public APIs. 
@baentsch

Reply:
:-( I won't respond to user questions about them.